pigpen-cascading/src/main/clojure/pigpen/cascading.clj

```diff
@@ -223,21 +243,26 @@
                                                            %) a))))))
          (remove is-field-projection))))
 
+
+(defn preprocess-commands [commands]
+  (-> commands
+      collapse-field-projections))
+
 (defn commands->flow
   "Transforms a series of commands into a Cascading flow"
-  [commands]
+  [commands ^FlowConnector connector]
   (clojure.pprint/pprint (preprocess-commands commands))
   (let [flowdef (reduce (fn [def cmd] (command->flowdef cmd def)) {} (preprocess-commands commands))
         {:keys [sources pipe-to-sink sinks pipes]} flowdef
         sources-map (into {} (map (fn [s] [(str s) (sources s)]) (keys sources)))
         sinks-map (into {} (map (fn [[p s]] [(str p) (sinks s)]) pipe-to-sink))
         tail-pipes (into-array Pipe (map #(pipes %) (keys pipe-to-sink)))]
-    (.connect (HadoopFlowConnector.) sources-map sinks-map tail-pipes)))
+    (.connect connector sources-map sinks-map tail-pipes)))
 
 (defn generate-flow
   "Transforms the relation specified into a Cascading flow that is ready to be executed."
-  ([query] (generate-flow {} query))
-  ([opts query]
+  ([query] (generate-flow (HadoopFlowConnector.) query))
+  ([connector query]
     (-> query
-        (oven/bake :cascading {} opts)
-        commands->flow)))
+        (oven/bake :cascading {} {})
```

Comment:
Take a look at `pigpen.pig.oven` - a similar approach might be a good place for the pre-processing that you're doing in here.